Generalize BCH and Reed-Solomon codes

mhostetter opened the pull request.

justinkang221 commented:
Great initiative. Would love to use this, as I might have to implement something like this myself soon. What is the state of things with your last commit?

mhostetter commented:
@justinkang221 I just rebased and pushed my latest changes. I haven't touched it for a bit now, so I'm not 100% refreshed on where I left off.

Here are some notes:
- The documentation *definitely* isn't complete, but some has been updated as I refactored. The docs built from this branch can be seen here: https://galois--413.org.readthedocs.build/en/413/api/#forward-error-correction.
- I'm still trying to "standardize" the structure of the parity-check matrix **H**.
- The unit tests aren't complete yet (and aren't all passing yet). I'm still testing / debugging and comparing to Sage.

The main purpose of this branch is to support general BCH and RS codes. This means primitive and non-primitive, narrow-sense and not, binary and q-ary BCH codes. Also, I'd like to support easier code construction -- either from the code size `(n, k)` or the code length `n` and design distance `d`,  eg `galois.BCH(n, k)` or `galois.BCH(n, d=d)`.

I welcome your testing and feedback on the branch. I cannot guarantee the code is bug-free. In fact, I know it's not. I think some modification to the decoder is still required. Feel free to post usage questions or observed bugs here in this thread.

mhostetter commented:
@justinkang221 I think this branch is ready for testing. I fixed a bug in the q-ary BCH decoder. Any feedback on usage, API, or discovered bugs would be very helpful.

The documentation hasn't been fully re-polished, but the `__init__` functions for BCH and RS are updated. You can find them here:
- https://galois--413.org.readthedocs.build/en/413/api/galois.BCH.__init__/
- https://galois--413.org.readthedocs.build/en/413/api/galois.ReedSolomon.__init__/
